module/builder/collection/builder.go

```diff
@@ -60,277 +53,244 @@ func NewBuilder(
+	if minPossibleRefHeight > refChainFinalizedHeight {
+		minPossibleRefHeight = 0 // overflow check
+	}
 
-		// TODO (ramtin): enable this again
-		// b.tracer.FinishSpan(parentID, trace.COLBuildOnSetup)
-		// b.tracer.StartSpan(parentID, trace.COLBuildOnUnfinalizedLookup)
-		// defer b.tracer.FinishSpan(parentID, trace.COLBuildOnUnfinalizedLookup)
-
-		// RATE LIMITING: the builder module can be configured to limit the
-		// rate at which transactions with a common payer are included in
-		// blocks. Depending on the configured limit, we either allow 1
-		// transaction every N sequential collections, or we allow K transactions
-		// per collection.
-
-		// keep track of transactions in the ancestry to avoid duplicates
-		lookup := newTransactionLookup()
-		// keep track of transactions to enforce rate limiting
-		limiter := newRateLimiter(b.config, parent.Height+1)
-
-		// look up previously included transactions in UN-FINALIZED ancestors
-		ancestorID := parentID
-		clusterFinalID := clusterFinal.ID()
-		for ancestorID != clusterFinalID {
-			ancestor, err := b.clusterHeaders.ByBlockID(ancestorID)
-			if err != nil {
-				return fmt.Errorf("could not get ancestor header (%x): %w", ancestorID, err)
-			}
-
-			if ancestor.Height <= clusterFinal.Height {
-				return fmt.Errorf("should always build on last finalized block")
-			}
-
-			payload, err := b.payloads.ByBlockID(ancestorID)
-			if err != nil {
-				return fmt.Errorf("could not get ancestor payload (%x): %w", ancestorID, err)
-			}
-
-			collection := payload.Collection
-			for _, tx := range collection.Transactions {
-				lookup.addUnfinalizedAncestor(tx.ID())
-				limiter.addAncestor(ancestor.Height, tx)
-			}
-			ancestorID = ancestor.ParentID
-		}
+	// TODO (ramtin): enable this again
+	// b.tracer.FinishSpan(parentID, trace.COLBuildOnSetup)
+	// b.tracer.StartSpan(parentID, trace.COLBuildOnUnfinalizedLookup)
+	// defer b.tracer.FinishSpan(parentID, trace.COLBuildOnUnfinalizedLookup)
+
+	// STEP TWO: create a lookup of all previously used transactions on the
+	// part of the chain we care about. We do this separately for
+	// un-finalized and finalized sections of the chain to decide whether to
+	// remove conflicting transactions from the mempool.
+
+	// keep track of transactions in the ancestry to avoid duplicates
+	lookup := newTransactionLookup()
+	// keep track of transactions to enforce rate limiting
+	limiter := newRateLimiter(b.config, parent.Height+1)
+
+	// RATE LIMITING: the builder module can be configured to limit the
+	// rate at which transactions with a common payer are included in
+	// blocks. Depending on the configured limit, we either allow 1
+	// transaction every N sequential collections, or we allow K transactions
+	// per collection.
+
+	// first, look up previously included transactions in UN-FINALIZED ancestors
+	fmt.Println("cluster chain final height: ", clusterChainFinalizedBlock.Height, clusterChainFinalizedBlock.ID())
```

Comment:
do we need this log?

Reply:
If we do want to log, it's better to log with zerolog. 

It actually makes sense to me that the builder to have a logger module, and be able to log things.

The builder right now does a lot of work and takes a lot of time to build a block. 

It would be good to log some useful metrics to understand which part is the bottleneck. 

For instance, I would be curious to know who many cluster blocks in total did builder actually iterate through to build the lookup. And see if that makes a big impact to block building time.